### PR TITLE
Improve error message for DecomposeLinearExpression and DecomposeAffineExpression.

### PR DESCRIPTION
**Pull request by hongkai-dai**
Motivated by slack discussion https://drakedevelopers.slack.com/archives/C3L92BM2Q/p1692903961875319

This change is [Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake/20065)

### CODE REVIEW

**Review by hongkai-dai — COMMENTED** [review]
This is

+@ggould-tri for feature review please since you just touched this code, thanks!

> *[Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake/20065#-:-NcexOGY89gQ62m98qXq:b-omvv83)* status: LGTM missing from assignee ggould-tri(platform), needs at least two assigned reviewers, missing label for release notes (waiting on @hongkai-dai)

**Review by ggould-tri — COMMENTED** [review]
:lgtm:; one typo, one wording suggestion.

> Reviewed 1 of 1 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake/20065#-:-NcwKSuL9Kxa87p8j-__:br2ua1s)* status: 2 unresolved discussions, needs at least two assigned reviewers, missing label for release notes (waiting on @hongkai-dai)

___
*[`common/symbolic/decompose.cc` line 144 at r1](https://reviewable.io/reviews/RobotLocomotion/drake/20065#-NcwJ70x3naekefcWY36:-NcwJ70yB2AAZh9uSeZ_:b-7ilvfu) ([raw file](https://github.com/RobotLocomotion/drake/blob/ff1b65450334bf782de8019763ece3ad67ffa8bc/common/symbolic/decompose.cc#L144)):*
> ```cpp
>       ThrowError(
>           "non-linear", e.to_string(),
>           fmt::format(" of indeterminates{}, with a constant term  {}. "
> ```

typo?



_Suggestion:_
```cpp
          fmt::format(" of indeterminates {}, with a constant term {}. "
```
___
*[`common/symbolic/decompose.cc` line 147 at r1](https://reviewable.io/reviews/RobotLocomotion/drake/20065#-NcwJGJ66FkJlHsO7qnS:-NcwJGJ66FkJlHsO7qnT:bece59a) ([raw file](https://github.com/RobotLocomotion/drake/blob/ff1b65450334bf782de8019763ece3ad67ffa8bc/common/symbolic/decompose.cc#L147)):*
> ```cpp
>                       "Linear expression "
>                       "should have no constant term, but affine expression has",
>                       fmt_eigen(vars.transpose()), map.at(Monomial{})));
> ```

FYI:
The text emitted here (after the typo fix above) will look something like,
`While decomposing an expression, we detected a non-linear expression: x+2 of indeterminates x, with a constant term 2. Linear expression should have no constant term, but affine expression has.`
The last sentence seems a bit confusingly worded to me.  I'd suggest rewording it more like,
"This is an affine expression; a linear expression should have no constant term."
Author's discretion, though -- you're the expert on these matters.
___
*[`common/symbolic/decompose.cc` line 626 at r1](https://reviewable.io/reviews/RobotLocomotion/drake/20065#-NcwLNzG997MX1IiCYkJ:-NcwLNzG997MX1IiCYkK:b-m1jwnw) ([raw file](https://github.com/RobotLocomotion/drake/blob/ff1b65450334bf782de8019763ece3ad67ffa8bc/common/symbolic/decompose.cc#L626)):*
> ```cpp
>   VectorX<Expression> w0(f.size());
>   for (int i = 0; i < f.size(); i++) {
>     auto const [w, alpha, this_w0] =
> ```

BTW:  Your linter is objecting to this.  IMO the specific linter gripe about whitespace is wrong, but it's wrong because const and auto are usually written in the other order:
```
ggould@baymax:~/git/drake$ git grep 'const *\[' | wc -l
1
ggould@baymax:~/git/drake$ git grep 'const auto *\[' | wc -l
96
```



_Suggestion:_
```cpp
    const auto [w, alpha, this_w0] =
```

**Review by hongkai-dai — COMMENTED** [review]
+@rpoyner-tri for Tuesday's platform review please, thanks!

> *[Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake/20065#-:-Ncxv23Q1VqWkkix8Zub:b6x4hq0)* status: LGTM missing from assignee rpoyner-tri(platform), missing label for release notes (waiting on @hongkai-dai)

**Review by ggould-tri — COMMENTED** [review]
> Reviewed 1 of 1 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake/20065)* status: LGTM missing from assignee rpoyner-tri(platform), missing label for release notes (waiting on @hongkai-dai)

**Review by rpoyner-tri — COMMENTED** [review]
:lgtm:

> Reviewed 1 of 1 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake/20065#-:-Nd0zzBYAOnu5BYIFlih:bnfp4nl)* status: missing label for release notes (waiting on @hongkai-dai)
